core/common/src/geometry/CoordinateReferenceSystem.ts

```
@@ -0,0 +1,457 @@
+  datumId?: string;
+  /** The complete definition of the geodetic datum referred to by datumId. It can also be used if the datum is not stored
+   *  in either service or dictionary.
+   */
+  datum?: GeodeticDatumProps;
+  /** The identifier of the geodetic ellipsoid as stored in the dictionary or the service database. This property is exclusive
+   * of having datumId and datum properties undefined.
+  */
+  ellipsoidId?: string;
+  /** The complete definition of the geodetic datum referred to by datumId. It can also be used if the datum is not stored
+   *  in either service or dictionary
+   */
+  ellipsoid?: GeodeticEllipsoidProps;
+  unit?: UnitType;
+  projection?: ProjectionProps;
+  area?: HorizontalCRSAreaProps;
+}
+
+/** Horizontal Geographic Coordinate reference System implementation.
+ *  An horizontal CRS defines the portion which is horizontal to the Earth surface (within the deformation brought by the projection process).
+ *  There are two major classes of Horizontal Coordinate Reference Systems:
+ *  - The projected CRS which rely on a projection to flatten the coordinate system space into axises using linear
+ *    units (meter, US Survey Feet, ...) relative to some origin.
+ *  - The non projected CRS (also named geographic CRS by EPSG nomenclature) that does
+ *    not require a projection (projection method = None) and horizontal coordinates are expressed
+ *    as longitude and latitude (see [[Cartographic]])
+ *  Horizontal Coordinate Systems rely on a projection to flatten the surface of an ellipsoid (see [[GeodeticEllipsoid]]) which is
+ *  the mathematical model of the Earth surface.
+ *  Explanations in more details of the various concepts can be obtained from other sources including the page on the subject
+ *  on itwinJS.org.
+ *  A few details are still required to grasp the model. Geographic Coordinate Reference Systems rely on the concept of geodetic datums
```

Did you consider making this class immutable to avoid all the awkwardness around property setters that silently fail based on state of other properties?

---

All classes have been made immutable by setting all properties readonly. I am a novice in TypeScript and could not find a syntax to make the whole class immutable. If there is a cleaner way, please advise.
Awkward setter and getters have been removed.